### PR TITLE
Refactor cirunners

### PR DESCRIPTION
**Pull request by b4ldr**
this is a rebase of #147.  i know this is a lot of code and likely too much to get accepted in its current form.  however i wanted to pull in your changes so had to go through the rebase regardless and thought its still something to leave here as food for thought.  also for the record i actually only ended up using the `shell_runner` myself so more then happy to switch to a different data structre

### CODE REVIEW

**Comment by b4ldr** [comment]
happy to clean up ruby syntax if there is interest in this

**Comment by ashish1099** [comment]
any update on this.

**Comment by eputnam** [comment]
@b4ldr any progress on this?

**Comment by b4ldr** [comment]
@eputnam im waiting for =feedback on the project authors to see if they want this PR or if they want me to re-work it.  Im currently using this code as is so nothing for me to do unless i get some feedback

File: manifests/cirunner/docker.pp [hunk 1]
```diff
@@ -0,0 +1,70 @@
+# == Class: gitlab::cirunner::docker
+#
+class gitlab::cirunner::docker (
+  Boolean                    $manage_docker                     = true,
+  String                     $xz_package_name                   = 'xz-utils',
+  String                     $default_image                     = 'ubuntu_trusty',
```
Comment (LongLiveCHIEF) [inline manifests/cirunner/docker.pp]: I think it might be better to make something a bit leaner the default image. Perhaps `docker`? Not going to make this a review comment, just a discussion I want to start.

**Comment by b4ldr** [comment]
hi @LongLiveCHIEF 

thanks for the comments and more then happy to discuss changing defaults and structure.  

>My biggest question; why did you decide to create a separate manifest for the various executors, >instead of using the runner executor as a parameter for something like a general cirunner config >manifest?

so by this i assume you mean why have `runner\{ssh,docker,etc}.pp` vs trying to implement everything in `runner.pp` if so then the simple reason is that the arguments passed to the register command are so different.  compare `runner::docker_ssh` vs `runner:: kubernetes`.  I looked to see if there was a way to inherit in puppet so theire wasn't so much copy and paste but couldn't find anything.  

I'm also not the biggest fan of the structure here myself but couldn't think of a better way to do it.  If you can i would be happy to work on refactoring.

**Comment by LongLiveCHIEF** [comment]
yeah, i def have some magic up my sleeve for this. I'll put some psuedocode down for you tonight to look over and consider.

**Comment by LongLiveCHIEF** [comment]
@b4ldr i've been looking at some psuedocode for an hour now. I def have some improvements, but they're just basic improvements to the multiple manifests, and not anything that reduces the number of manifests for executors.

This... is not easy. 

Did you put any thought into developing the cirunner manifests around creating a `config.toml` template, perhaps similar to how we handle the `gitlab.rb` configuration?  What did you find (if you considered it)?

**Comment by b4ldr** [comment]
@LongLiveCHIEF i have not explored using the confi.toml but i think you main end up with the same problem

**Comment by LongLiveCHIEF** [comment]
Ok. I'm still working through this, and I expect my review will take about another week or so. I make some extensive use of this feature, so testing should be quite thorough. I'll have myself or anyone on my team comment here if we have questions throughout the week.

**Comment by b4ldr** [comment]
ok thanks i will wait untill your review is over before rebasing

**Comment by bastelfreak** [comment]
@b4ldr thanks for this awesome PR. There are quite a few changes in this single commit that are not related to the new feature. Can you please pull them out into seperate PRs? This make reviewing way way easier. Also it would allow us to revert single commits in the future if one of them introduces a regression.

File: manifests/cirunner.pp [hunk 2]
```diff
@@ -34,52 +34,36 @@
...
-  $concurrent = undef,
-  $metrics_server = undef,
-  $hiera_default_config_key = 'gitlab_ci_runners_defaults',
-  $hiera_runners_key = 'gitlab_ci_runners',
-  $manage_docker = true,
-  $manage_repo = true,
-  $xz_package_name = 'xz-utils',
-  $package_ensure = installed,
-  $package_name = 'gitlab-runner',
+  Optional[Integer]       $concurrent           = undef,
```
Comment (bastelfreak) [inline manifests/cirunner.pp]: datatypes \o/

----

I really love them, but adding them to existing params should be a dedicated PR

File: manifests/cirunner.pp [hunk 3]
```diff
@@ -34,52 +34,36 @@
...
+  Hash                    $docker_runners       = {},
+  Hash                    $shell_runners        = {},
+  Hash                    $ssh_runners          = {},
+  Hash                    $docker_ssh_runners   = {},
+  Hash                    $parallels_runners    = {},
+  Hash                    $virtualbox_runners   = {},
+  Hash                    $kubernetes_runners   = {},
 ) {
 
-  validate_string($hiera_default_config_key)
```
Comment (bastelfreak) [inline manifests/cirunner.pp]: removing those validate calls should be made with the datatypes for old params in another PR.

File: manifests/cirunner.pp [hunk 4]
```diff
@@ -34,52 +34,36 @@
...
-    }
-  }
-
   if $manage_repo {
     $repo_base_url = 'https://packages.gitlab.com'
 
     case $::osfamily {
       'Debian': {
-        include apt
+        include ::apt
```
Comment (bastelfreak) [inline manifests/cirunner.pp]: Why was this changed? using topscope isn't needed for Puppet4 and newer, in our review guidelines we agreed on not doing this. https://voxpupuli.org/docs/#reviewing-a-module-pr

File: manifests/cirunner.pp [hunk 5]
```diff
@@ -165,14 +147,53 @@
...
-  gitlab::runner { $runners:
-    binary         => $package_name,
-    default_config => $default_config,
-    runners_hash   => $runners_hash,
-    require        => Exec['gitlab-runner-restart'],
+  if ! empty($docker_runners) {
+    if defined(Class['gitlab::cirunner::docker']) {
+      warning('Class[\'gitlab::cirunner::docker\' is defined so $gitlab::cirunner::docker__runners willbe ignored')
+    } else {
+      class {'::gitlab::cirunner::docker': runners => $docker_runners}
```
Comment (bastelfreak) [inline manifests/cirunner.pp]: you don't need the :: in front of the class names

File: manifests/cirunner/docker.pp [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  Hash                       $docker_images                     = {},
+  Hash                       $runners                           = {},
+  Optional[String]           $default_token                     = undef,
+  Optional[String]           $default_url                       = undef,
+  Optional[Boolean]          $default_run_untagged              = undef,
```
Comment (bastelfreak) [inline manifests/cirunner/docker.pp]: is it possible that the booleans are not Optional but default to true/false? That often makes sense (not always) and is then the preferred solution.

File: manifests/runner/docker.pp [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+    $tags_cmd, $host_cmd, $tlsverify_cmd, $hostname_cmd, $image_cmd,
+    $cpuset_cpus_cmd, $cpus_cmd, $dns_cmd, $dns_search_cmd, $privileged_cmd,
+    $userns_cmd, $cap_add_cmd, $cap_drop_cmd, $security_opt_cmd, $devices_cmd,
+    $disable_cache_cmd, $volume_driver_cmd, $extra_hosts_cmd, $volumes_from_cmd,
+    $network_mode_cmd, $links_cmd, $services_cmd, $wait_for_services_timeout_cmd,
+    $allowed_images_cmd, $allowed_services_cmd, $pull_policy_cmd, $shm_size_cmd,
+  ].join(' ')
+  # Execute gitlab ci multirunner register
+  exec {"Register_runner_${title}":
+    command => "/usr/bin/gitlab-ci-multi-runner register ${parameters_string}",
```
Comment (bastelfreak) [inline manifests/runner/docker.pp]: should we use https://github.com/puppetlabs/puppetlabs-stdlib#shell_escape or https://github.com/puppetlabs/puppetlabs-stdlib#shell_join here?

Comment (b4ldr) [reply manifests/runner/docker.pp]: I could probably replace the join used above with the shell_join function.  however this variable has no user submitted data so it should be safe

File: manifests/cirunner.pp (hunk 2 above)
Comment (b4ldr) [reply manifests/cirunner.pp]: cirunner.pp is a complete re-write so im not sure a dedicated PR makes sense

File: manifests/cirunner.pp (hunk 3 above)
Comment (b4ldr) [reply manifests/cirunner.pp]: cirunner.pp is a complete re-write so im not sure a dedicated PR makes sense

File: manifests/cirunner.pp (hunk 4 above)
Comment (b4ldr) [reply manifests/cirunner.pp]: reverted back

File: manifests/cirunner.pp (hunk 5 above)
Comment (b4ldr) [reply manifests/cirunner.pp]: done

File: manifests/cirunner/docker.pp (hunk 6 above)
Comment (b4ldr) [reply manifests/cirunner/docker.pp]: would need to double check but i think . making them all false would help be fine

File: manifests/runner/docker.pp (hunk 7 above)
Comment (b4ldr) [reply manifests/runner/docker.pp]: i can change to shell_join i think

File: manifests/cirunner/docker.pp (hunk 6 above)
Comment (LongLiveCHIEF) [reply manifests/cirunner/docker.pp]: these are correct as is. Please don't change them.

File: lib/puppet/functions/gitlab/cirunner/cmd_str.rb [hunk 8]
```diff
@@ -0,0 +1,18 @@
+Puppet::Functions.create_function('gitlab::cirunner::cmd_str') do
+  dispatch :cmd_str do
+    param 'String', :name
+    param 'Array', :values
+    # only avalible in puppet 4.7
```
Comment (alexjfisher) [inline lib/puppet/functions/gitlab/cirunner/cmd_str.rb]: I think Voxpupuli’s minimum supported puppet version is 4.10, so you should be fine setting the return type.

File: manifests/runner/docker.pp (hunk 7 above)
Comment (b4ldr) [reply manifests/runner/docker.pp]: @bastelfreak i looked at using both shell_join and shell_escape however neither of theses produced the desired command.

### shell_join produced
`/usr/bin/gitlab-ci-multi-runner register --executor\\ docker --registration-token\\ TOKEN --url\\ https://gitlab.com -n\\ --name\\ docker_docker_runner --run-untagged '' '' --docker-host\\ foobar '' '' --docker-image\\ ubuntu_trusty '' '' '' '' '' '' '' '' '' '' '' '' '' '' '' '' '' '' '' '' '' ''`

### join(' ').shell_escape produced
`/usr/bin/gitlab-ci-multi-runner register --executor\\ docker\\ --registration-token\\ TOKEN\\ --url\\ https://gitlab.com\\ -n\\ --name\\ docker_docker_runner\\ --run-untagged\\ \\ \\ --docker-host\\ foobar\\ \\ \\ --docker-image\\ ubuntu_trusty\\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\`

**Comment by b4ldr** [comment]
is there still any interest in this?

**Comment by LongLiveCHIEF** [comment]
Yes, there is.  I just took over as the lead maintainer on this, and I'm taking another look at it this week.  There are some cleanup things I'd like to see, some rebasing that's needed, and I'm still wondering if there's a way this can be broken up into multiple PR's?

**Comment by b4ldr** [comment]
@LongLiveCHIEF its a while since i worked on this code but happy to continue it if there is interest, just let me know how you want things to look.

**Comment by LongLiveCHIEF** [comment]
After some thought, I think this should be its own puppet module.  The module currently handles installation and configuration of the gitlab-omnibus package, and there is absolutely no strict dependency between the functionality and gitlab omnibus.

**Comment by LongLiveCHIEF** [comment]
@b4ldr I created https://github.com/voxpupuli/puppet-gitlab-ci-runner. I'm a heavy user of `gitlab-puppet`, and willing to continue to maintain it, but after the first release of `puppet-gitlab-ci-runner`, i'm not sure how much I'll want to stay involved with it.

Would you like to be added as a collaborator for that project, and help me release the first release and maintain it moving forward?

**Comment by LongLiveCHIEF** [comment]
I'm going to be migrating all relevant open issues, and relevant closed PR's from this repo over to that repository as well, so this will get re-opened there as I get everything extracted.

**Comment by b4ldr** [comment]
@LongLiveCHIEF yes please add me and ill assist best i can

**Comment by bastelfreak** [comment]
Hi @b4ldr, I added you to test list of collaborators. We've got a reviewguideline online at https://voxpupuli.org/docs/#reviewing-a-module-pr. If you've got any questions don't hesitate to ping us on http://puppetcommunity.slack.com or our IRC channel #voxpupuli on freenode.
